### PR TITLE
curl: update to 7.88.1

### PR DESCRIPTION
**Pull request by stangri**
Maintainer: me
Compile tested: x86_64, Sophos XG-105w, OpenWrt 22.03.3
Run tested: x86_64, Sophos XG-105w, OpenWrt 22.03.3, use with https-dns-proxy

Description:
* https://curl.se/changes.html#7_88_1

### CODE REVIEW

**Comment by stangri** [comment]
@trippleflux can you test this with `perl-www-curl`?

**Comment by BKPepe** [comment]
I still wonder, where did you take the "1" in the commit subject as versioning, when the version is 7.88.1 even homepage says that: https://curl.se/download.html

**Comment by trippleflux** [comment]
@stangri, compiling right now, will edit this comment once compilation test has finished.

**Comment by stangri** [comment]
> @stangri, compiling right now, will edit this comment once compilation test has finished.

Please add a new comment instead so I'd receive an email notification. ;)

**Comment by trippleflux** [comment]
@stangri 
Hmm, failures : 
```
/home/username/works/openwrt/staging_dir/target-x86_64_glibc_custom/usr/include/curl/curl.h:2515:3: note: declared here
 2515 |   CURLFORM_CONTENTTYPE     CURL_DEPRECATED(7.56.0, "Use curl_mime_type()"),
      |   ^~~~~~~~~~~~~~~~~~~~
make[3]: *** [Makefile:347: Curl.o] Error 1
make[3]: Leaving directory '/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/perl/WWW-Curl-4.17'
make[2]: *** [Makefile:52: /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/perl/WWW-Curl-4.17/.built] Error 2
make[2]: Leaving directory '/home/username/works/openwrt/feeds/packages/lang/perl-www-curl'
time: package/feeds/packages/perl-www-curl/compile#1.32#0.96#2.22
    ERROR: package/feeds/packages/perl-www-curl failed to build.
```

**Comment by trippleflux** [comment]
If adding the following into perl-www-curl Makefile : 
```
MOD_CFLAGS_PERL += -w
```

The error goes as below : 
```
"/home/username/works/openwrt/staging_dir/hostpkg/usr/bin/perl5.28.1" "-Iinc" -MExtUtils::Command::MM -e 'cp_nonempty' -- Curl.bs blib/arch/auto/WWW/Curl/Curl.bs 644
"/home/username/works/openwrt/staging_dir/hostpkg/usr/bin/perl5.28.1" "-Iinc" "/home/username/works/openwrt/staging_dir/hostpkg/usr/lib/perl5/5.28.1/ExtUtils/xsubpp"  -typemap '/home/username/works/openwrt/staging_dir/hostpkg/usr/lib/perl5/5.28.1/ExtUtils/typemap' -typemap '/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/perl/WWW-Curl-4.17/typemap'  Curl.xs > Curl.xsc
mv Curl.xsc Curl.c
x86_64-openwrt-linux-gcc -c  -I/home/username/works/openwrt/staging_dir/target-x86_64_glibc_custom/usr/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fno-caller-saves -fno-plt -mtune=tigerlake -march=tigerlake -O3 -funsafe-math-optimizations -ftree-vectorize -fhonour-copts -fmacro-prefix-map=/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/perl/WWW-Curl-4.17=WWW-Curl-4.17 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/usr/include -I/home/username/works/openwrt/staging_dir/toolchain-x86_64_gcc-12.2.0_glibc/include -D_REENTRANT -D_GNU_SOURCE -w -O2   -DVERSION=\"4.17\" -DXS_VERSION=\"4.17\" -fPIC "-I/home/username/works/openwrt/staging_dir/target-x86_64_glibc_custom/usr/lib/perl5/5.28/CORE/"   Curl.c
In file included from Curl.xs:574:
curlopt-constants.c: In function 'constant':
curlopt-constants.c:19:51: error: 'CURL_DEPRECATED' undeclared (first use in this function)
   19 |             if (strEQ(name, "DEPRECATED")) return CURL_DEPRECATED;
      |                                                   ^~~~~~~~~~~~~~~
curlopt-constants.c:19:51: note: each undeclared identifier is reported only once for each function it appears in
curlopt-constants.c:66:59: error: 'CURL_IGNORE_DEPRECATION' undeclared (first use in this function)
   66 |             if (strEQ(name, "IGNORE_DEPRECATION")) return CURL_IGNORE_DEPRECATION;
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~~
curlopt-constants.c:2809:58: error: 'CURLOPTDEPRECATED' undeclared (first use in this function); did you mean 'CURLOPT_PRIVATE'?
 2809 |             if (strEQ(name, "CURLOPTDEPRECATED")) return CURLOPTDEPRECATED;
      |                                                          ^~~~~~~~~~~~~~~~~
      |                                                          CURLOPT_PRIVATE
make[3]: *** [Makefile:347: Curl.o] Error 1
make[3]: Leaving directory '/home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/perl/WWW-Curl-4.17'
make[2]: *** [Makefile:54: /home/username/works/openwrt/build_dir/target-x86_64_glibc_custom/perl/WWW-Curl-4.17/.built] Error 2
make[2]: Leaving directory '/home/username/works/openwrt/feeds/packages/lang/perl-www-curl'
time: package/feeds/packages/perl-www-curl/compile#1.24#0.36#1.54
    ERROR: package/feeds/packages/perl-www-curl failed to build.
make[1]: *** [package/Makefile:116: package/feeds/packages/perl-www-curl/compile] Error 1
make[1]: Leaving directory '/home/username/works/openwrt'
make: *** [/home/username/works/openwrt/include/toplevel.mk:231: package/perl-www-curl/compile] Error 2
```

**Comment by trippleflux** [comment]
Found the fixes for perl-www-curl : 
230-curl-7.87_compat.patch
```
diff -Naur a/Makefile.PL b/Makefile.PL
--- a/Makefile.PL
+++ b/Makefile.PL
@@ -127,7 +127,7 @@
     close H;
 
     for my $e (sort @syms) {
-       if($e =~ /(OBSOLETE|^CURL_EXTERN|CURLOPT\z|^CURL_STRICTER\z|_LAST\z|_LASTENTRY\z|WIN32)/) {
+       if($e =~ /(OBSOLETE|^CURL_EXTERN|^CURL_STRICTER\z|^CURL_DID_MEMORY_FUNC_TYPEDEFS\z|_LAST\z|_LASTENTRY\z|^CURLINC_|^CURL_WIN32\z|^CURL_DEPRECATED\z|^CURL_IGNORE_DEPRECATION\z|^CURLOPTDEPRECATED\z|^CURLOPT\z)/) {
           next;
        }
        my ($group) = $e =~ m/^([^_]+_)/;
```

Haven't tested if functionalities wise perl-www-curl is working fine, but compiled fine for x86_64, feel free to submit a PR based on my finding.

**Comment by stangri** [comment]
> Found the fixes for perl-www-curl : 230-curl-7.87_compat.patch
> 
> ```
> diff -Naur a/Makefile.PL b/Makefile.PL
> --- a/Makefile.PL
> +++ b/Makefile.PL
> @@ -127,7 +127,7 @@
>      close H;
>  
>      for my $e (sort @syms) {
> -       if($e =~ /(OBSOLETE|^CURL_EXTERN|CURLOPT\z|^CURL_STRICTER\z|_LAST\z|_LASTENTRY\z|WIN32)/) {
> +       if($e =~ /(OBSOLETE|^CURL_EXTERN|^CURL_STRICTER\z|^CURL_DID_MEMORY_FUNC_TYPEDEFS\z|_LAST\z|_LASTENTRY\z|^CURLINC_|^CURL_WIN32\z|^CURL_DEPRECATED\z|^CURL_IGNORE_DEPRECATION\z|^CURLOPTDEPRECATED\z|^CURLOPT\z)/) {
>            next;
>         }
>         my ($group) = $e =~ m/^([^_]+_)/;
> ```
> 
> Haven't tested if functionalities wise perl-www-curl is working fine, but compiled fine for x86_64, feel free to submit a PR based on my finding.

@Naoir @ryzhovau @ecsv @champtar do you gentlemen have any interest in creating a PR with the above?

**Comment by BKPepe** [comment]
I'm thinking that you can create pull request as well, right? :innocent: In that case, you will not rely on others until they would be able to create it.

**Comment by champtar** [comment]
Does this curl update break perl-www-curl ? If yes why not add the patch in this PR in a second commit ?

**Comment by stangri** [comment]
> Does this curl update break perl-www-curl ? If yes why not add the patch in this PR in a second commit ?

I can't test it.

**Comment by BKPepe** [comment]
You can try to compile it, thats the thing.

Dne čt 23. 2. 2023 5:07 uživatel Stan Grishin ***@***.***>
napsal:

**Comment by stangri** [comment]
@BKPepe contribution guidelines require a run-test, that I can't do.

If between you and @champtar you consider compile-test sufficient I can add another commit with the fix for `perl-www-curl`, please advise.

**Comment by champtar** [comment]
@stangri you can make everything compile, and have @trippleflux run test, team work is definitely accepted

**Comment by trippleflux** [comment]
@BKPepe & @stangri 
I am sorry that I cannot directly make a PR but don't feel like sharing my real name currently.

@champtar 
That's the other problematic side from me, I don't have much or more perl programming backgound to conform the run test. If you have perl snippet codes to run on to conforming the perl-www-curl functionalities would be nice but it needs my real name to signing the run-tested?

**Comment by BKPepe** [comment]
Is there any specific reason why did you put the patch for perl in your commit to have it together? It's just confusing. From my point of view, I think you are trying to reinvent the wheel, which is not necessary as someone did it for you in the past. 

If you take a look at the history: https://github.com/openwrt/packages/commits/master/lang/perl-www-curl/Makefile

You see that it the patch should be in a separate commit to have good, understandable Git history, which you can use to find things.

**Comment by BKPepe** [comment]
Uh, I wonder what you were trying to do... 
- You were adding the patch to ``perl-www-curl``, why did you choose ``curl`` as the prefix for your commit subject? 
  - As you already might know, it is better to put some references in the pull request. Otherwise, with each force push, you are spamming it. 

Even I gave you a hint of how it should look. :shrug: 

TL;DR: I've fixed it for you how it should look and pushed it to your repository.
